File: script/make/protocol/lua/RankCenterProtocol.lua
RankCenterProtocol = {}

function RankCenterProtocol.encode(offset, protocol, data)
    if protocol == 19101 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19102 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19103 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19104 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19105 then
        local offset = offset
        local table = {}
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function RankCenterProtocol.decode(offset, protocol, data)
    if protocol == 19101 then
        local offset = offset
        -- 排行榜
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Rank
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId}
            list[listIndex] = rank
        end
        return {list = list}
    elseif protocol == 19102 then
        local offset = offset
        -- 排行榜
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Rank
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local level = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local classes = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = level, classes = classes}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            list[listIndex] = rank
        end
        return {list = list}
    elseif protocol == 19103 then
        local offset = offset
        -- 排行榜
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Rank
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local level = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local classes = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 性别
            local sex = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = level, classes = classes, sex = sex}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            list[listIndex] = rank
        end
        return {list = list}
    elseif protocol == 19104 then
        local offset = offset
        -- 排行榜
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Rank
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local level = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local classes = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 性别
            local sex = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- VIP等级
            local vipLevel = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = level, classes = classes, sex = sex, vipLevel = vipLevel}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            list[listIndex] = rank
        end
        return {list = list}
    elseif protocol == 19105 then
        local offset = offset
        -- 排行榜
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Rank
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local level = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local classes = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 性别
            local sex = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- VIP等级
            local vipLevel = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 头像
            local avatar = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = level, classes = classes, sex = sex, vipLevel = vipLevel, avatar = avatar}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            list[listIndex] = rank
        end
        return {list = list}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/js/RankCenterProtocol.js
export default class RankCenterProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 19101: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19102: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19103: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19104: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19105: {
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 19101: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // object
                    const rank = {type, order, key, value, time, name, serverId};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19102: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19103: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 性别
                    const sex = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes, sex};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19104: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 性别
                    const sex = view.getUint8(offset, false);
                    offset = offset + 1;
                    // VIP等级
                    const vipLevel = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes, sex, vipLevel};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19105: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 性别
                    const sex = view.getUint8(offset, false);
                    offset = offset + 1;
                    // VIP等级
                    const vipLevel = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 头像
                    const avatar = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes, sex, vipLevel, avatar};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}